oak_attestation_verification/src/util.rs

```diff
@@ -134,52 +134,37 @@ pub enum MatchResult {
     UNDECIDABLE = 2,
 }
 
-/// Compares two binary arrays byte per byte.
-fn match_strings(a: &str, b: &str) -> MatchResult {
-    if a != b {
-        return MatchResult::DIFFERENT;
-    }
-    MatchResult::SAME
+fn matches<T: Eq>(a: &T, b: &T) -> MatchResult {
+    if a == b { MatchResult::SAME } else { MatchResult::DIFFERENT }
+}
+
+// Generates a match set between two structurs `a` and `b` for every field `f`
+// provided. The fields will be checked in the order provided. The first pair of
+// non-empty fields will be compared. If no field is non-empty in both structs,
+// the result is undecidable.
+macro_rules! first_non_empty_match {
+    ($a:ident, $b:ident: $($f:ident),*) => {
+       match ($a,$b) {
+            $(
+                ($a, $b) if !$a.$f.is_empty() && !$b.$f.is_empty() => {
+                    matches(&$a.$f, &$b.$f)
+                }
+            )*
+            _ => MatchResult::UNDECIDABLE
+        }
+    };
 }
 
 /// Compares two hex digests.
 pub fn is_hex_digest_match(a: &HexDigest, b: &HexDigest) -> MatchResult {
-    if !a.psha2.is_empty() && !b.psha2.is_empty() {
-        return match_strings(&a.psha2, &b.psha2);
-    }
-    if !a.sha2_256.is_empty() && !b.sha2_256.is_empty() {
-        return match_strings(&a.sha2_256, &b.sha2_256);
-    }
-    if !a.sha2_512.is_empty() && !b.sha2_512.is_empty() {
-        return match_strings(&a.sha2_512, &b.sha2_512);
-    }
-    if !a.sha3_512.is_empty() && !b.sha3_512.is_empty() {
-        return match_strings(&a.sha3_512, &b.sha3_512);
-    }
-    if !a.sha3_384.is_empty() && !b.sha3_384.is_empty() {
-        return match_strings(&a.sha3_384, &b.sha3_384);
-    }
-    if !a.sha3_256.is_empty() && !b.sha3_256.is_empty() {
-        return match_strings(&a.sha3_256, &b.sha3_256);
-    }
-    if !a.sha3_224.is_empty() && !b.sha3_224.is_empty() {
-        return match_strings(&a.sha3_224, &b.sha3_224);
-    }
-    if !a.sha2_384.is_empty() && !b.sha2_384.is_empty() {
-        return match_strings(&a.sha2_384, &b.sha2_384);
-    }
-
-    // Nit: Put the weak hash to the end of comparisons.
-    if !a.sha1.is_empty() && !b.sha1.is_empty() {
-        return match_strings(&a.sha1, &b.sha1);
-    }
-
-    MatchResult::UNDECIDABLE
+    first_non_empty_match!(a, b:
```

Comment:
Thanks a lot for looking into this! FWIW in the meantime there is a fourth state "CONTRADICTORY" which makes this fully symmetric: //k/c/o/release/import_utils.py, line 92. This removes the dependency on the order which is awkward. As a downside there is no early return, all fields are checked beginning to end. I was planning on putting it here.

Reply:
Good to know!

On a somewhat related note, I wonder if it would make sense to change the return type to use Result?

For example, something like:

```
enum MatchError {
  UndecidableError,
  ContradictoryError,
}

Result<Boolean, MatchError>
```